gateway-js/src/index.ts

```diff
@@ -467,6 +445,102 @@ export class ApolloGateway implements GraphQLService {
     }
   }
 
+  private async updateCsdl(result: CsdlUpdate): Promise<void> {
+    if (result.id === this.compositionId) {
+      this.logger.debug('No change in composition since last check.');
+      return;
+    }
+
+    const previousSchema = this.schema;
+    const previousCsdl = this.parsedCsdl;
+    const previousCompositionId = this.compositionId;
+
+    if (previousSchema) {
+      this.logger.info('Updated CSDL was found.');
+    }
+
+    await this.maybePerformServiceHealthCheck(result);
+
+    this.compositionId = result.id;
+    this.parsedCsdl = parse(result.csdl);
```

Comment:
Can we do anything that might throw, such as this `parse` call, before we start changing the state at all? If this parse call throws then we've already updated compositionId without updating anything else, which seems bad. If we can manage to calculate all the stuff we would be putting on `this` fields first and then write all the fields at once that would be nice.

(Or we could even take all the fields that need to be replaced whenever you update the schema, put them on their own object (kinda like "schema derived data" in AS), and then swap that sub-object over with a single assignment when it's ready.)

Reply:
I think your suggestion is a great follow-up / cleanup PR. It will require a bit of unwinding due to fns calling fns which depend on `this` things being updated. It should be done carefully and separate from this PR.

However, I agree with gracefully handling throws from `parse` (now `gql`) - this should be resolved in af77271d with a couple test additions.

Reply:
I've taken what was previously the `parse` issue (#580) and repurposed it to capture the suggested follow-up here. Resolving this issue will both resolve the duplicate `parse` calls as well as simplify/standardize how the gateway updates its internals a bit as the update is happening.